src/timemachine-heap.c

```diff
@@ -0,0 +1,150 @@
+    if (unlikely(heap == NULL)) {
+        return NULL;     
+    }
+
+    heap->conf = conf;
+    heap->used_pool_count = 0;
+    heap->unused_pool_count = 0;
+    
+    TAILQ_INIT(&heap->used_mem_pools);
+    TAILQ_INIT(&heap->unused_mem_pools);
+    TAILQ_INIT(&heap->unused_packets);    
+    
+    uint32_t expanded = TimeMachineHeapExpand(tmtv, heap, conf->prealloc_count);
+    if (expanded != conf->prealloc_count) {
+        SCLogError(SC_ERR_MEM_ALLOC, "Error: Could not allocate initial memory for "
+                   "time machine, expanded heap by: %" PRIu32 ", requested: %" PRIu32 
+                   " current Memory: %" PRIu64 " max Memory: %" PRIu64, expanded, 
+                   conf->prealloc_count, tmtv->current_memory, timemachine_config.max_memory);
+        exit(EXIT_FAILURE);                         
+    }
+    
+    return heap;
+}
+
+void TimeMachineHeapDestroy(TimeMachineHeap* heap) {
+    
+    if (heap == NULL) {
+        return;
+    }
+    
+    /* delete all the used pool data */
+    while (heap->used_pool_count > 0) {
+        TimeMachineMemPool* mem_pool;
+            
+        mem_pool = TAILQ_FIRST(&heap->used_mem_pools);
+        TAILQ_REMOVE(&heap->used_mem_pools, mem_pool, next);
+        SCFree(mem_pool);
+
+        heap->used_pool_count--;    
+    }
+    
+    /* delete all the unused pool data */
+    while (heap->unused_pool_count > 0) {
+        TimeMachinePacket* packet;
+        TimeMachineMemPool* mem_pool;
+
+        packet = TAILQ_FIRST(&heap->unused_packets);
+        TAILQ_REMOVE(&heap->unused_packets, packet, next);
+        TimeMachinePacketDestroy(packet);
+            
+        mem_pool = TAILQ_FIRST(&heap->unused_mem_pools);
+        TAILQ_REMOVE(&heap->unused_mem_pools, mem_pool, next);
+        SCFree(mem_pool);
+                
+        heap->unused_pool_count--;
+    }    
+}
+
+uint32_t TimeMachineHeapExpand(TimeMachineThreadVars* tmtv, TimeMachineHeap* heap, uint32_t expand_by) {
+    uint32_t i;
+           
+    if (heap == NULL) {
+        return 0;
+    }
+
+    uint32_t expand_count = expand_by;
+    uint32_t expand_mem_size = sizeof(TimeMachineMemPool) + heap->conf->max_packet_size + 
+                      sizeof(TimeMachinePacket);
+                   
+    if ((expand_mem_size * expand_by) + tmtv->current_memory > timemachine_config.max_memory) {
+        expand_count = (timemachine_config.max_memory - tmtv->current_memory) / expand_mem_size + 1;
+    }    
+    
+    for (i = 0; i < expand_count; i++) {
+        TimeMachineMemPool* mem_pool;
+        TimeMachinePacket* packet;
+        
+        mem_pool = SCCalloc(sizeof(TimeMachineMemPool), 1);
+        if (mem_pool == NULL) {
+            return i;
+        }    
+        
+        mem_pool->mem = SCMalloc(heap->conf->max_packet_size);
+        if (mem_pool->mem == NULL) {
+            SCFree(mem_pool);
+            return i;
+        }
+        
+        packet = TimeMachinePacketNew();
```

Comment:
I'm a bit worried about the thread safety here.

The flow engine adds packets to the heap in it's own TimeMachineThreadVars. As this is done in the 'private' thread data, no locking or other synchronisation is done.

Writing out of packet data is done in AlertPcapLogProcess. It walks the time machine packets stored in the flow
<pre>
TimeMachinePacket* packet = TAILQ_FIRST(&p->flow->tm_pkts);
</pre>
and through packet->data references the mempool that is part of the heap in the TimeMachineThreadVars.

Correct so far?

The problem is that this may not be the same thread.

In the autofp runmode the flow engine part is done in the capture thread(s). It's unclear to me where AlertPcapLogProcess would run currently, but its correct place would be in the OuputPacketModule API as a suggested here https://github.com/inliniac/suricata/pull/1817#discussion_r51350700 That would mean multiple output threads would all potentially reference the data in the capture thread(s).

What do you think?